### PR TITLE
[SPARK-43113][SQL] Evaluate stream-side variables when generating code for a bound condition

### PR DESCRIPTION
**Pull request by bersprockets**
### What changes were proposed in this pull request?

In `JoinCodegenSupport#getJoinCondition`, evaluate any referenced stream-side variables before using them in the generated code.

This patch doesn't evaluate the passed stream-side variables directly, but instead evaluates a copy (`streamVars2`). This is because `SortMergeJoin#codegenFullOuter` will want to evaluate the stream-side vars within a different scope than the condition check, so we mustn't delete the initialization code from the original `ExprCode` instances.

### Why are the changes needed?

When a bound condition of a full outer join references the same stream-side column more than once, wholestage codegen generates bad code.

For example, the following query fails with a compilation error:

```
create or replace temp view v1 as
select * from values
(1, 1),
(2, 2),
(3, 1)
as v1(key, value);

create or replace temp view v2 as
select * from values
(1, 22, 22),
(3, -1, -1),
(7, null, null)
as v2(a, b, c);

select *
from v1
full outer join v2
on key = a
and value > b
and value > c;
```
The error is:
```
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 277, Column 9: Redefinition of local variable "smj_isNull_7"
```
The same error occurs with code generated from ShuffleHashJoinExec:
```
select /*+ SHUFFLE_HASH(v2) */ *
from v1
full outer join v2
on key = a
and value > b
and value > c;
```
In this case, the error is:
```
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 174, Column 5: Redefinition of local variable "shj_value_1" 
```
Neither `SortMergeJoin#codegenFullOuter` nor `ShuffledHashJoinExec#doProduce` evaluate the stream-side variables before calling `consumeFullOuterJoinRow#getJoinCondition`. As a result, `getJoinCondition` generates definition/initialization code for each referenced stream-side variable at the point of use. If a stream-side variable is used more than once in the bound condition, the definition/initialization code is generated more than once, resulting in the "Redefinition of local variable" error.

In the end, the query succeeds, since Spark disables wholestage codegen and tries again.

(In the case other join-type/strategy pairs, either the implementations don't call `JoinCodegenSupport#getJoinCondition`, or the stream-side variables are pre-evaluated before the call is made, so no error happens in those cases).

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New unit tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @rednaxelafx FYI

**Review by HyukjinKwon — APPROVED** [review]
Looks making sense to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/JoinCodegenSupport.scala [hunk 1]
```diff
@@ -42,13 +42,15 @@ trait JoinCodegenSupport extends CodegenSupport with BaseJoinExec {
       buildRow: Option[String] = None): (String, String, Seq[ExprCode]) = {
     val buildSideRow = buildRow.getOrElse(ctx.freshName("buildRow"))
     val buildVars = genOneSideJoinVars(ctx, buildSideRow, buildPlan, setDefaultValue = false)
+    val streamVars2 = streamVars.map(_.copy())
     val checkCondition = if (condition.isDefined) {
       val expr = condition.get
       // evaluate the variables from build side that used by condition
-      val eval = evaluateRequiredVariables(buildPlan.output, buildVars, expr.references)
+      val eval = evaluateRequiredVariables(buildPlan.output ++ streamPlan.output,
```
Comment (bersprockets) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/JoinCodegenSupport.scala:49]: You know, for consistency with code below I should really make this `streamPlan.output ++ buildPlan.output` and `streamVars2 ++ buildVars`.

Comment (bersprockets) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/JoinCodegenSupport.scala:49]: Also I should update the comment above this line.

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.4 and branch-3.3.

**Comment by HyukjinKwon** [comment]
It has some minor conflicts against branch-3.3. @bersprockets wanna make a pr for it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/JoinCodegenSupport.scala [hunk 2]
```diff
@@ -42,13 +42,15 @@ trait JoinCodegenSupport extends CodegenSupport with BaseJoinExec {
       buildRow: Option[String] = None): (String, String, Seq[ExprCode]) = {
     val buildSideRow = buildRow.getOrElse(ctx.freshName("buildRow"))
     val buildVars = genOneSideJoinVars(ctx, buildSideRow, buildPlan, setDefaultValue = false)
+    val streamVars2 = streamVars.map(_.copy())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/JoinCodegenSupport.scala:45]: late LGTM. Can we add some comments here? This is very tricky and we should tell others why a copy is needed. The problem is `ExprCode` is mutable, `streamVars` may be used again by full outer joins and we should not change its states here.

**Pull request closed without merging**